jaxopt/_src/tree_util.py

```diff
@@ -78,38 +78,35 @@ def tree_zeros_like(tree_x):
   return tree_map(jnp.zeros_like, tree_x)
 
 
-def tree_collapse(tree):
-  """Takes a pytree of arrays in input and return 1D vector of flattened leaves.
-
-  Reciprocal function of ``tree_uncollapse``.
+def tree_average(trees, weights):
+  """Return the linear combination of a list of trees.
 
   Args:
-    tree: tree of arrays
+    trees: tree of arrays with shape (m,...)
+    weights: array of shape (m,)
+
   Returns:
-    a 1D vector
+    a single tree that is the linear combination of all trees
   """
-  leaves = tree_leaves(tree)
-  leaves = list(map(lambda params: params.flatten(), leaves))
-  return jnp.concatenate(leaves, axis=0)
+  return tree_map(lambda x: jnp.tensordot(weights, x, axes=1), trees)
 
 
-def tree_uncollapse(tree, vec):
-  """Returns a pytree with same elements as 1D vector ``vec`` and with same structure as ``tree``.
+def pytree_at_set(tree, pos, values):
+  at_set = lambda x, value: x.at[pos].set(value)
+  return tree_map(at_set, tree, values)
 
-  Reciprocal function of ``tree_collapse``. Transform the 1D vector into a pytree
-  compatible with ``tree`` structure, with same shapes.
+
+def tree_batch_vdot(a, b):
```

Comment:
Doesn't seem to be used anywhere -> delete.

Reply:
I use it to build the Gram matrix in history. I renamed it to reflect its true semantic.